packages/flutter_tools/lib/src/commands/upgrade.dart

```diff
@@ -29,64 +38,182 @@ class UpgradeCommand extends FlutterCommand {
 
   @override
   Future<FlutterCommandResult> runCommand() async {
+    final UpgradeCommandRunner upgradeCommandRunner = UpgradeCommandRunner();
+    await upgradeCommandRunner.runCommand(argResults['force'], GitTagVersion.determine(), FlutterVersion.instance);
+    return null;
+  }
+}
+
+
+@visibleForTesting
+class UpgradeCommandRunner {
+  Future<FlutterCommandResult> runCommand(bool force, GitTagVersion gitTagVersion, FlutterVersion flutterVersion) async {
+    await verifyUpstreamConfigured();
+    if (!force && gitTagVersion == const GitTagVersion.unknown()) {
+      // If the commit is a recognized branch and not master,
+      // explain that we are avoiding potential damage.
+      if (flutterVersion.channel != 'master' && FlutterVersion.officialChannels.contains(flutterVersion.channel)) {
+        throwToolExit(
+          'Unknown flutter tag. Abandoning upgrade to avoid destroying local '
+          'changes. It is recommended to use git directly if not working off of '
+          'an official channel.'
+        );
+      // Otherwise explain that local changes can be lost.
+      } else {
+        throwToolExit(
+          'Unknown flutter tag. Abandoning upgrade to avoid destroying local '
+          'changes. If it is okay to remove local changes, then re-run this '
+          'command with --force.'
+        );
+      }
+    }
+    final String stashName = await maybeStash(gitTagVersion);
+    await upgradeChannel(flutterVersion);
+    await attemptRebase();
+    await precacheArtifacts();
+    await updatePackages(flutterVersion);
+    await runDoctor();
+    await applyStash(stashName);
```

Comment:
you probably want to apply the stash before you try to precache artifacts, otherwise if the stash changes the tool in some way it won't have taken effect in time.